Access denied ("java.lang.RuntimePermission" "accessClassInPackage.sun.nio.cs.ext")

@aguibert mentioned a problem running Liberty on OpenJ9 Java 9

```
>Stack Dump = java.lang.ExceptionInInitializerError
>    at java.lang.J9VMInternals.ensureError(java.base@9-internal/J9VMInternals.java:191)
>    at java.lang.J9VMInternals.recordInitializationFailure(java.base@9-internal/J9VMInternals.java:180)
>    at java.nio.charset.Charset.lookupExtendedCharset(java.base@9-internal/Charset.java:456)
>    at java.nio.charset.Charset.lookup2(java.base@9-internal/Charset.java:486)
>    at java.nio.charset.Charset.lookup(java.base@9-internal/Charset.java:474)
>    at java.nio.charset.Charset.forName(java.base@9-internal/Charset.java:538)
>    at com.sun.org.apache.xml.internal.serializer.Encodings$EncodingInfos.findCharsetNameFor(java.xml@9-internal/Encodings.java:402)
>    at com.sun.org.apache.xml.internal.serializer.Encodings$EncodingInfos.findCharsetNameFor(java.xml@9-internal/Encodings.java:438)
>    at com.sun.org.apache.xml.internal.serializer.Encodings$EncodingInfos.loadEncodingInfo(java.xml@9-internal/Encodings.java:466)
>    at com.sun.org.apache.xml.internal.serializer.Encodings$EncodingInfos.<init>(java.xml@9-internal/Encodings.java:324)
>    at com.sun.org.apache.xml.internal.serializer.Encodings$EncodingInfos.<init>(java.xml@9-internal/Encodings.java:312)
>    at com.sun.org.apache.xml.internal.serializer.Encodings.<clinit>(java.xml@9-internal/Encodings.java:580)
>    at com.sun.org.apache.xml.internal.serializer.ToStream.<init>(java.xml@9-internal/ToStream.java:135)
>    at com.sun.org.apache.xml.internal.serializer.ToXMLStream.<init>(java.xml@9-internal/ToXMLStream.java:65)
>    at com.sun.org.apache.xalan.internal.xsltc.runtime.output.TransletOutputHandlerFactory.getSerializationHandler(java.xml@9-internal/TransletOutputHandlerFactory.java:163)
>    at com.sun.org.apache.xalan.internal.xsltc.trax.TransformerImpl.getOutputHandler(java.xml@9-internal/TransformerImpl.java:472)
>    at com.sun.org.apache.xalan.internal.xsltc.trax.TransformerImpl.transform(java.xml@9-internal/TransformerImpl.java:361)
>    at com.ibm.ws.webcontainer.osgi.mbeans.PluginGenerator.generateXML(PluginGenerator.java:756)
>    at com.ibm.ws.webcontainer.osgi.mbeans.GeneratePluginConfigMBean.generatePluginConfig(GeneratePluginConfigMBean.java:159)
>    at com.ibm.ws.webcontainer.osgi.mbeans.GeneratePluginConfigMBean.generatePluginConfig(GeneratePluginConfigMBean.java:261)
>    at com.ibm.ws.webserver.plugin.runtime.listeners.GeneratePluginConfigListener$1.run(GeneratePluginConfigListener.java:285)
>    at java.util.concurrent.Executors$RunnableAdapter.call(java.base@9-internal/Executors.java:514)
>    at java.util.concurrent.FutureTask.run(java.base@9-internal/FutureTask.java:264)
>    at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@9-internal/ThreadPoolExecutor.java:1167)
>    at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@9-internal/ThreadPoolExecutor.java:641)
>    at java.lang.Thread.run(java.base@9-internal/Thread.java:835)
>Caused by: java.security.AccessControlException: Access denied ("java.lang.RuntimePermission" "accessClassInPackage.sun.nio.cs.ext")
>    at java.security.AccessController.throwACE(java.base@9-internal/AccessController.java:176)
>    at java.security.AccessController.checkPermissionHelper(java.base@9-internal/AccessController.java:236)
>    at java.security.AccessController.checkPermission(java.base@9-internal/AccessController.java:371)
>    at java.lang.SecurityManager.checkPermission(java.base@9-internal/SecurityManager.java:558)
>    at java.lang.SecurityManager.checkPackageAccess(java.base@9-internal/SecurityManager.java:1534)
>    at java.lang.J9VMInternals$2.run(java.base@9-internal/J9VMInternals.java:311)
>    at java.security.AccessController.doPrivileged(java.base@9-internal/AccessController.java:664)
>    at java.lang.J9VMInternals.checkPackageAccess(java.base@9-internal/J9VMInternals.java:307)
>    at java.lang.ClassLoader.defineClassImpl(java.base@9-internal/Native Method)
>    at java.lang.ClassLoader.defineClass(java.base@9-internal/ClassLoader.java:445)
>    at java.lang.ClassLoader.defineClass(java.base@9-internal/ClassLoader.java:1985)
>    at java.security.SecureClassLoader.defineClass(java.base@9-internal/SecureClassLoader.java:206)
>    at jdk.internal.loader.BuiltinClassLoader.defineClass(java.base@9-internal/BuiltinClassLoader.java:1021)
>    at jdk.internal.loader.BuiltinClassLoader.lambda$findClassInModuleOrNull$4(java.base@9-internal/BuiltinClassLoader.java:865)
>    at jdk.internal.loader.BuiltinClassLoader$$Lambda$155.00000000B0AFBFA0.run(Unknown Source)
>    at java.security.AccessController.doPrivileged(java.base@9-internal/AccessController.java:638)
>    at jdk.internal.loader.BuiltinClassLoader.findClassInModuleOrNull(java.base@9-internal/BuiltinClassLoader.java:866)
>    at jdk.internal.loader.BuiltinClassLoader.findClass(java.base@9-internal/BuiltinClassLoader.java:733)
>    at java.lang.ClassLoader.loadClassHelper(java.base@9-internal/ClassLoader.java:1131)
>    at java.lang.Class.forName(java.base@9-internal/Class.java:442)
>    at java.util.ServiceLoader.lambda$loadProvider$1(java.base@9-internal/ServiceLoader.java:852)
>    at java.util.ServiceLoader$$Lambda$154.00000000B0AF9BC0.run(Unknown Source)
>    at java.security.AccessController.doPrivileged(java.base@9-internal/AccessController.java:694)
>    at java.util.ServiceLoader.loadProvider(java.base@9-internal/ServiceLoader.java:854)
>    at java.util.ServiceLoader.access$600(java.base@9-internal/ServiceLoader.java:390)
>    at java.util.ServiceLoader$ModuleServicesLookupIterator.hasNext(java.base@9-internal/ServiceLoader.java:1071)
>    at java.util.ServiceLoader$2.hasNext(java.base@9-internal/ServiceLoader.java:1294)
>    at java.util.ServiceLoader$3.hasNext(java.base@9-internal/ServiceLoader.java:1379)
>    at java.nio.charset.Charset$ExtendedProviderHolder$1.run(java.base@9-internal/Charset.java:442)
>    at java.nio.charset.Charset$ExtendedProviderHolder$1.run(java.base@9-internal/Charset.java:436)
>    at java.security.AccessController.doPrivileged(java.base@9-internal/AccessController.java:638)
>    at java.nio.charset.Charset$ExtendedProviderHolder.extendedProviders(java.base@9-internal/Charset.java:436)
>    at java.nio.charset.Charset$ExtendedProviderHolder.<clinit>(java.base@9-internal/Charset.java:433)
```

--- a/runtime/oti/vm_api.h
+++ b/runtime/oti/vm_api.h
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 1991, 2017 IBM Corp. and others
+ * Copyright (c) 1991, 2018 IBM Corp. and others
  *
  * This program and the accompanying materials are made available under
  * the terms of the Eclipse Public License 2.0 which accompanies this
@@ -2215,6 +2215,19 @@ fieldIndexTableRemove(J9JavaVM* vm, J9Class *ramClass);
  */
 UDATA   
 packageAccessIsLegal(J9VMThread *currentThread, J9Class *targetClass, j9object_t protectionDomain, UDATA canRunJavaCode);
+
+/**
+ * Determine if the targetClass is subject to the package access check.
+ * Assuming targetClass is NOT NULL.
+ *
+ * @param *vm[in] the J9JavaVM
+ * @param srcClassLoader[in] Current class loader creating RAM class
+ * @param srcModule[in] The module where current RAM class is loaded from
+ * @param targetClass[in] The RAM class being examined to be determined if a package access check should be performed
+ * @return TRUE if a package access check is required, otherwise FALSE.
+ */
+BOOLEAN
+requirePackageAccessCheck(J9JavaVM *vm, J9ClassLoader *srcClassLoader, J9Module *srcModule, J9Class *targetClass);
 
 /**
  * @brief

--- a/runtime/vm/createramclass.cpp
+++ b/runtime/vm/createramclass.cpp
@@ -142,7 +142,7 @@ static UDATA checkPackageAccess(J9VMThread *vmThread, J9Class *foundClass, UDATA
 static void setCurrentExceptionForBadClass(J9VMThread *vmThread, J9UTF8 *badClassName, UDATA exceptionIndex);
 static BOOLEAN verifyClassLoadingStack(J9VMThread *vmThread, J9ClassLoader *classLoader, J9ROMClass *romClass);
 static void popFromClassLoadingStack(J9VMThread *vmThread);
-static VMINLINE BOOLEAN loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9ROMClass *romClass, J9Class *elementClass, UDATA packageID, BOOLEAN hotswapping, UDATA classPreloadFlags, J9Class **superclassOut);
+static VMINLINE BOOLEAN loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9ROMClass *romClass, J9Class *elementClass, UDATA packageID, BOOLEAN hotswapping, UDATA classPreloadFlags, J9Class **superclassOut, J9Module *module);
 #if defined(J9VM_OPT_VALHALLA_NESTMATES)
 static J9Class *loadNestHost(J9VMThread *vmThread, J9ClassLoader *classLoader, J9UTF8 *nestHostName, UDATA classPreloadFlags);
 #endif /* defined(J9VM_OPT_VALHALLA_NESTMATES) */
@@ -1495,7 +1495,6 @@ popFromClassLoadingStack(J9VMThread *vmThread)
 	pool_removeElement(vmThread->javaVM->classLoadingStackPool, topOfStack);
 }
 
-
 /**
  * Attempts to recursively load (if necessary) the required superclass and
  * interfaces for the class being loaded.
@@ -1507,12 +1506,13 @@ popFromClassLoadingStack(J9VMThread *vmThread)
  */
 static VMINLINE BOOLEAN
 loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9ROMClass *romClass, J9Class *elementClass,
-	UDATA packageID, BOOLEAN hotswapping, UDATA classPreloadFlags, J9Class **superclassOut)
+	UDATA packageID, BOOLEAN hotswapping, UDATA classPreloadFlags, J9Class **superclassOut, J9Module *module)
 {
+	J9JavaVM *vm = vmThread->javaVM;
 	const BOOLEAN isROMClassUnsafe = (J9ROMCLASS_IS_UNSAFE(romClass) != 0);
 	J9UTF8 *className = J9ROMCLASS_CLASSNAME(romClass);
-	J9UTF8 *superclassName;
-	J9Class *superclass;
+	J9UTF8 *superclassName = NULL;
+	J9Class *superclass = NULL;
 
 	superclassName = J9ROMCLASS_SUPERCLASSNAME(romClass);
 	if (superclassName == NULL) {
@@ -1538,7 +1538,9 @@ loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9
 		}
 
 		if (!hotswapping) {
-			if (checkPackageAccess(vmThread, superclass, classPreloadFlags) != 0) {
+			if (requirePackageAccessCheck(vm, classLoader, module, superclass) 
+				&& (checkPackageAccess(vmThread, superclass, classPreloadFlags) != 0)
+			) {
 				return FALSE;
 			}
 
@@ -1578,7 +1580,9 @@ loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9
 					if (interfaceClass == NULL) {
 						return FALSE;
 					}
-					if (checkPackageAccess(vmThread, interfaceClass, classPreloadFlags) != 0) {
+					if (requirePackageAccessCheck(vm, classLoader, module, interfaceClass) 
+						&& (checkPackageAccess(vmThread, interfaceClass, classPreloadFlags) != 0)
+					) {
 						return FALSE;
 					}
 					/* ensure that the interface is in fact an interface */
@@ -2714,14 +2718,14 @@ internalCreateRAMClassFromROMClass(J9VMThread *vmThread, J9ClassLoader *classLoa
 	IDATA entryIndex, I_32 locationType, J9Class *classBeingRedefined, J9Class *hostClass)
 {
 	J9JavaVM *javaVM = vmThread->javaVM;
-	J9Class *superclass;
-	J9UTF8 *className;
-	UDATA packageID;
+	J9Class *superclass = NULL;
+	J9UTF8 *className = NULL;
+	UDATA packageID = 0;
 	UDATA classPreloadFlags = 0;
 	BOOLEAN hotswapping = (0 != (options & J9_FINDCLASS_FLAG_NO_DEBUG_EVENTS));
 	BOOLEAN fastHCR = (0 != (options & J9_FINDCLASS_FLAG_FAST_HCR));
-	J9CreateRAMClassState state;
-	J9Class *result;
+	J9CreateRAMClassState state = {0};
+	J9Class *result = NULL;
 	J9ClassLoader* hostClassLoader = classLoader;
 	J9Module* module = NULL;
 
@@ -2822,7 +2826,7 @@ retry:
 		}
 	}
 
-	if (!loadSuperClassAndInterfaces(vmThread, hostClassLoader, romClass, elementClass, packageID, hotswapping, classPreloadFlags, &superclass)) {
+	if (!loadSuperClassAndInterfaces(vmThread, hostClassLoader, romClass, elementClass, packageID, hotswapping, classPreloadFlags, &superclass, module)) {
 		omrthread_monitor_enter(javaVM->classTableMutex);
 		return internalCreateRAMClassDone(vmThread, classLoader, romClass, options, elementClass, className, &state);
 	}

--- a/runtime/vm/resolvesupport.cpp
+++ b/runtime/vm/resolvesupport.cpp
@@ -20,6 +20,7 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
  *******************************************************************************/
 
+#include "j2sever.h"
 #include "j9.h"
 #include "j9protos.h"
 #include "j9consts.h"
@@ -120,6 +121,25 @@ packageAccessIsLegal(J9VMThread *currentThread, J9Class *targetClass, j9object_t
 		}
 	}
 	return legal;
+}
+
+BOOLEAN
+requirePackageAccessCheck(J9JavaVM *vm, J9ClassLoader *srcClassLoader, J9Module *srcModule, J9Class *targetClass)
+{
+	BOOLEAN checkFlag = TRUE;
+	if (J2SE_VERSION(vm) >= J2SE_19) {
+		if (srcModule == targetClass->module) {
+			if (NULL != srcModule) {
+				/* same named module */
+				checkFlag = FALSE;
+			} else if (srcClassLoader == targetClass->classLoader) {
+				/* same unnamed module */
+				checkFlag = FALSE;
+			}
+		}
+	}
+	
+	return checkFlag;
 }
 
 j9object_t   
@@ -359,8 +379,12 @@ tryAgain:
 	 * No check is required if any of the following is true:
 	 * 		- the current class and resolved class are identical
 	 * 		- the current class was loaded by the bootstrap class loader
+	 * 		- the current class and resolved class are in same module
 	 */
-	if ((currentClass != resolvedClass) && (classLoader != bootstrapClassLoader)) {
+	if ((currentClass != resolvedClass) 
+		&& (classLoader != bootstrapClassLoader)
+		&& requirePackageAccessCheck(vm, classLoader, currentClass->module, resolvedClass)
+	) {
 		/* AOT resolves class refs inside J9Classes which have not yet
 		 * had the java/lang/Class associated with them.  canRunJavaCode must be false
 		 * in this case.  The protectionDomain object is only used if canRunJavaCode

